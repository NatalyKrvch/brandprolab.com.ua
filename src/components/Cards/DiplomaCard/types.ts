export interface DiplomaCardProps {
  diplomaURL: string;
  onClick?: () => void;
}
