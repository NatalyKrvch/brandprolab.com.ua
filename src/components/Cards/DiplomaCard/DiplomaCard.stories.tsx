import type { Meta, StoryObj } from '@storybook/react';

import DiplomaCard from './DiplomaCard';

const meta: Meta<typeof DiplomaCard> = {
  title: 'Components/DiplomaCard',
  component: DiplomaCard,
  parameters: {
    controls: { expanded: true },
  },
  argTypes: {
    diplomaURL: {
      control: 'text',
      description: 'The URL of the diploma image.',
      table: {
        type: { summary: 'string' },
      },
    },
    onClick: {
      action: 'clicked',
      description: 'Handler triggered when the diploma is clicked.',
      table: {
        type: { summary: '() => void' },
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof DiplomaCard>;

export const Vertical: Story = {
  args: {
    diplomaURL:
      'https://media.vanityfair.com/photos/60073e65d507f0e0031b84c5/master/w_2560%2Cc_limit/VF0321_Cover_SUB-OPENER.png',
  },
};

export const Horizontal: Story = {
  args: {
    diplomaURL:
      'https://media.vanityfair.com/photos/63e561bc4d9ea7ac2d152348/16:9/w_5791,h_3257,c_limit/VU0323_Cover_EdLetterPage.jpg',
  },
};
