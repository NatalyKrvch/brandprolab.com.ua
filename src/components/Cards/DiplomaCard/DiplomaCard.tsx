import Image from 'next/image';

import {
  DIPLOMA_CARD_TEST_ID,
  DIPLOMA_IMAGE_TEST_ID,
  DIPLOMA_OVERLAY_TEST_ID,
} from '@/lib/testIDs';

import { DiplomaCardProps } from './types';

const DiplomaCard = ({ diplomaURL, onClick }: DiplomaCardProps) => {
  return (
    <div
      className="w-fit rounded-24 border border-gray-medium p-2"
      data-testid={DIPLOMA_CARD_TEST_ID}
    >
      <div className="relative h-108 w-auto max-w-full overflow-hidden rounded-3xl">
        <div
          onClick={onClick}
          onKeyDown={e => e.key === 'Enter' && onClick?.()}
          className="absolute inset-0 z-10 bg-teal-fog opacity-20 transition-opacity duration-300 hover:cursor-pointer hover:opacity-0"
          data-testid={DIPLOMA_OVERLAY_TEST_ID}
        ></div>

        <Image
          src={diplomaURL}
          role="button"
          tabIndex={0}
          alt="Diploma"
          width={168}
          height={168}
          className="h-full w-auto object-contain"
          loading="lazy"
          data-testid={DIPLOMA_IMAGE_TEST_ID}
        />
      </div>
    </div>
  );
};

export default DiplomaCard;
