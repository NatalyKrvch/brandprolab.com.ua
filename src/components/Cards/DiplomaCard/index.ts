export { default as DiplomaCard } from './DiplomaCard';
export * from './types';
