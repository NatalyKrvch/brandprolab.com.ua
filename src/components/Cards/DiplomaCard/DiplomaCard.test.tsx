import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import {
  DIPLOMA_CARD_TEST_ID,
  DIPLOMA_IMAGE_TEST_ID,
  DIPLOMA_OVERLAY_TEST_ID,
} from '@/lib/testIDs';

import DiplomaCard from './DiplomaCard';

jest.mock('next/image', () => {
  const MockedImage = ({
    src,
    alt,
    className,
  }: {
    src: string;
    alt: string;
    className?: string;
  }) => (
    <img
      src={src}
      alt={alt}
      className={className}
      data-testid={DIPLOMA_IMAGE_TEST_ID}
    />
  );
  MockedImage.displayName = 'MockedNextImage';
  return MockedImage;
});

describe('DiplomaCard component', () => {
  const defaultProps = {
    diplomaURL: '/test-diploma.jpg',
    onClick: jest.fn(),
  };

  it('renders the DiplomaCard component', () => {
    const { asFragment } = render(<DiplomaCard {...defaultProps} />);
    const cardElement = screen.getByTestId(DIPLOMA_CARD_TEST_ID);

    expect(cardElement).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the diploma image correctly', () => {
    const { asFragment } = render(<DiplomaCard {...defaultProps} />);
    const imageElement = screen.getByTestId(DIPLOMA_IMAGE_TEST_ID);

    expect(imageElement).toBeInTheDocument();
    expect(imageElement).toHaveAttribute('src', '/test-diploma.jpg');
    expect(imageElement).toHaveAttribute('alt', 'Diploma');
    expect(asFragment()).toMatchSnapshot();
  });

  it('triggers onClick when overlay is clicked', () => {
    const { asFragment } = render(<DiplomaCard {...defaultProps} />);
    const overlayElement = screen.getByTestId(DIPLOMA_OVERLAY_TEST_ID);

    fireEvent.click(overlayElement);
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
