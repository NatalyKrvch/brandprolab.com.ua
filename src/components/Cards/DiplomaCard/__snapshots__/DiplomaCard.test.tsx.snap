// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiplomaCard renders correctly with black & white image and overlay 1`] = `
<div>
  <div
    class="hover:transition-color w-fit overflow-hidden rounded-24 border border-gray-medium p-3 hover:border-teal-dark hover:duration-500"
    data-testid="diploma-card-test-id"
  >
    <div
      class="relative h-230 w-auto max-w-full overflow-hidden rounded-2xl"
    >
      <div
        class="absolute inset-0 z-10 overflow-hidden bg-teal-fog opacity-20 transition-opacity duration-300 hover:cursor-pointer hover:opacity-30 hover:transition-opacity hover:duration-300"
        data-testid="diploma-overlay-test-id"
        role="button"
        tabindex="0"
      />
      <img
        alt="Diploma"
        class="h-full w-auto object-contain"
        data-nimg="1"
        data-testid="diploma-image-test-id"
        decoding="async"
        height="168"
        loading="lazy"
        src="/_next/image?url=https%3A%2F%2Fexample.com%2Fbw-diploma.jpg&w=384&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fexample.com%2Fbw-diploma.jpg&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fexample.com%2Fbw-diploma.jpg&w=384&q=75 2x"
        style="color: transparent;"
        width="168"
      />
    </div>
  </div>
</div>
`;
